File: react-native/src/network/TokenRefreshHandler.ts
/**
 * TokenRefreshHandler.ts
 * Manages token refresh with a mutex to prevent concurrent refresh calls.
 *
 * Strategy:
 * 1. POST /api/user/get_new_access_token/ with {refresh_token}
 * 2. On failure, POST /api/user/send_tokens/ with {device_id, user_id} and API-Key header
 *    This re-issues tokens for the SAME guest user — conversation history is preserved.
 *    (Do NOT fall back to initialize_user — that creates a brand new user, losing all history.)
 */

import { API_REFRESH_TOKEN, API_SEND_TOKENS, GUEST_API_KEY } from '../config/constants';
import { RefreshTokenRequest } from '../models/requests';
import { RefreshTokenResponse } from '../models/responses';
import { TokenExpiredError } from './NetworkError';
import { TokenStorage } from '../storage/TokenStorage';

// ─── Mutex ───────────────────────────────────────────────────────────────────

let refreshPromise: Promise<string> | null = null;

// ─── Public ──────────────────────────────────────────────────────────────────

/**
 * Obtains a new access token.
 * All concurrent callers share a single in-flight promise to avoid duplicate requests.
 */
export async function refreshIfNeeded(baseUrl: string): Promise<string> {
  if (refreshPromise) {
    return refreshPromise;
  }

  refreshPromise = (async (): Promise<string> => {
    try {
      return await attemptPrimaryRefresh(baseUrl);
    } catch {
      // Primary refresh failed — re-issue tokens for the same guest user via send_tokens
      try {
        return await attemptSendTokens(baseUrl);
      } catch {
        throw new TokenExpiredError(API_REFRESH_TOKEN);
      }
    }
  })();

  try {
    return await refreshPromise;
  } finally {
    refreshPromise = null;
  }
}

// ─── Step 1: normal token refresh ────────────────────────────────────────────

async function attemptPrimaryRefresh(baseUrl: string): Promise<string> {
  const refreshToken = await TokenStorage.getRefreshToken();
  if (!refreshToken) throw new Error('No refresh token stored');

  const url = `${baseUrl}/${API_REFRESH_TOKEN}`;
  const body: RefreshTokenRequest = { refresh_token: refreshToken };

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
    body: JSON.stringify(body),
  });

  if (!response.ok) {
    throw new Error(`Primary refresh failed: ${response.status}`);
  }

  const data: RefreshTokenResponse = await response.json();
  if (!data.access_token) throw new Error('No access_token in refresh response');

  await TokenStorage.saveTokens(
    data.access_token,
    data.refresh_token ?? refreshToken,
  );
  return data.access_token;
}

// ─── Step 2: re-issue tokens for the same guest user via send_tokens ─────────
//
// send_tokens uses the existing device_id + user_id to obtain new tokens for
// the SAME guest account. This preserves conversation history.
// Never fall back to initialize_user here — that creates a new user identity.

async function attemptSendTokens(baseUrl: string): Promise<string> {
  const deviceId = await TokenStorage.getOrCreateDeviceId();
  const userId = await TokenStorage.getUserId();
  if (!userId) throw new Error('No user_id stored — cannot call send_tokens');

  const url = `${baseUrl.replace(/\/$/, '')}/${API_SEND_TOKENS}`;

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      'API-Key': GUEST_API_KEY,
    },
    body: JSON.stringify({ device_id: deviceId, user_id: userId }),
  });

  if (!response.ok) {
    throw new Error(`send_tokens failed: ${response.status}`);
  }

  const data: RefreshTokenResponse = await response.json();
  if (!data.access_token) throw new Error('No access_token in send_tokens response');

  await TokenStorage.saveTokens(
    data.access_token,
    data.refresh_token ?? (await TokenStorage.getRefreshToken()) ?? '',
  );
  return data.access_token;
}

File: ios/Sources/FarmerChatSDK/Network/TokenRefreshHandler.swift
import Foundation

// MARK: - TokenRefreshHandler

/// Thread-safe actor that ensures only one token refresh runs at a time.
/// Additional callers are queued and receive the same result.
actor TokenRefreshHandler {

    // MARK: Private State

    private var isRefreshing = false
    private var waiters: [CheckedContinuation<String, Error>] = []

    // MARK: Public API

    /// Refresh the access token.  If a refresh is already in-flight, this
    /// method suspends until the existing refresh completes and then returns
    /// the same new token to every caller.
    ///
    /// - Parameters:
    ///   - baseUrl: Backend base URL.
    ///   - tokenStore: Store to read/write tokens.
    /// - Returns: The new access token.
    func refreshToken(baseUrl: String, tokenStore: any TokenStore) async throws -> String {
        // Queue behind an in-progress refresh
        if isRefreshing {
            return try await withCheckedThrowingContinuation { continuation in
                waiters.append(continuation)
            }
        }

        isRefreshing = true

        do {
            let newToken = try await performRefresh(baseUrl: baseUrl, tokenStore: tokenStore)
            resumeWaiters(with: .success(newToken))
            isRefreshing = false
            return newToken
        } catch {
            resumeWaiters(with: .failure(error))
            isRefreshing = false
            throw error
        }
    }

    // MARK: Private

    private func resumeWaiters(with result: Result<String, Error>) {
        let captured = waiters
        waiters.removeAll()
        for continuation in captured {
            continuation.resume(with: result)
        }
    }

    private func performRefresh(baseUrl: String, tokenStore: any TokenStore) async throws -> String {
        // Step 1: Try normal token refresh
        do {
            return try await attemptRefreshToken(baseUrl: baseUrl, tokenStore: tokenStore)
        } catch {
            // Step 2: Fall back to send_tokens (guest token acquisition)
            return try await attemptSendTokens(baseUrl: baseUrl, tokenStore: tokenStore)
        }
    }

    // MARK: Step 1 – /api/user/get_new_access_token/

    private func attemptRefreshToken(baseUrl: String, tokenStore: any TokenStore) async throws -> String {
        guard let refreshToken = tokenStore.getRefreshToken(), !refreshToken.isEmpty else {
            throw NetworkError.unauthorized
        }

        let urlString = baseUrl.trimmingCharacters(in: .init(charactersIn: "/"))
            + "/" + APIEndpoints.tokenRefresh

        guard let url = URL(string: urlString) else { throw NetworkError.invalidURL }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["refresh_token": refreshToken]
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw NetworkError.unknown("No HTTP response from token refresh")
        }

        guard httpResponse.statusCode == 200 else {
            throw NetworkError.serverError(httpResponse.statusCode, "Token refresh failed")
        }

        let tokenResponse = try JSONDecoder().decode(TokenResponse.self, from: data)

        guard let newAccess = tokenResponse.accessToken, !newAccess.isEmpty else {
            throw NetworkError.unknown("Empty access token in refresh response")
        }

        tokenStore.saveTokens(
            accessToken: newAccess,
            refreshToken: tokenResponse.refreshToken ?? refreshToken
        )
        return newAccess
    }

    // MARK: Step 2 – /api/user/send_tokens/

    private static let guestApiKey = "Y2K3kW5R9uQ0fL2X8zI7hT3aJ7"

    private func attemptSendTokens(baseUrl: String, tokenStore: any TokenStore) async throws -> String {
        guard let deviceId = tokenStore.getDeviceId(), !deviceId.isEmpty else {
            throw NetworkError.unauthorized
        }
        guard let userId = tokenStore.getUserId(), !userId.isEmpty else {
            throw NetworkError.unauthorized
        }

        let urlString = baseUrl.trimmingCharacters(in: .init(charactersIn: "/"))
            + "/" + APIEndpoints.sendTokens

        guard let url = URL(string: urlString) else { throw NetworkError.invalidURL }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(TokenRefreshHandler.guestApiKey, forHTTPHeaderField: "API-Key")

        let body: [String: String] = ["device_id": deviceId, "user_id": userId]
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw NetworkError.unknown("No HTTP response from send_tokens")
        }

        guard httpResponse.statusCode == 200 else {
            throw NetworkError.unauthorized
        }

        let tokenResponse = try JSONDecoder().decode(TokenResponse.self, from: data)

        guard let newAccess = tokenResponse.accessToken, !newAccess.isEmpty else {
            throw NetworkError.unauthorized
        }

        tokenStore.saveTokens(
            accessToken: newAccess,
            refreshToken: tokenResponse.refreshToken
        )
        return newAccess
    }
}
